dns: Fix DNS option `ndots`, `timeout` and `attempts`

cathay4t opened the pull request:
## Patch  1
The `ndots`, `timeout` and `attempts` DNS options are allowed to
hold a integer value in the format of `<opt_name>:<int>`. Previously,
nmstate is treating them as invalid DNS option. This patch fix it.

Now this YAML is supported:

```yml
dns-resolver:
  config:
    options:
    - rotate
    - ndots:9
```



## Patch 2

When user desires:

```yml
---
dns-resolver:
  config:
    search: []
```

It means user want to remove all search but preserve servers and
options, current nmstate incorrectly treat this as purge also.

This patch only treat these two as purge.

```yml
dns-resolver:
  config: {}
```

and

```yml
dns-resolver:
  config:
    server: []
    search: []
    options: []
```

cathay4t committed:
dns: Fix DNS option `ndots`, `timeout` and `attempts`

The `ndots`, `timeout` and `attempts` DNS options are allowed to
hold a integer value in the format of `<opt_name>:<int>`. Previously,
nmstate is treating them as invalid DNS option. This patch fix it.

Now this YAML is supported:

```yml
dns-resolver:
  config:
    options:
    - rotate
    - ndots:9
```

Unit and integration test cases included.

Signed-off-by: Gris Ge <fge@redhat.com>

liangwen12year commented on tests/integration/dns_test.py:
```diff
@@ -573,3 +573,39 @@ def test_invalid_dns_option_with_value(static_dns):
+
+    current_state = libnmstate.show()
+    assert DNS.OPTIONS not in current_state[DNS.KEY][DNS.CONFIG]
+    assert current_state[DNS.KEY][DNS.CONFIG][DNS.SERVER]
```
maybe instead of asserting the existence of DNS server and search, can we assert that they are actually preserved by adding a variable holding the state before applying the DNS option, then assert the old DNS server and search equal to the one after applying.

cathay4t replied:
Test case updated with assert on old server and search not changed.

cathay4t committed:
CI: Cancel on PR force push

Signed-off-by: Gris Ge <fge@redhat.com>

cathay4t committed:
dns: Fix purging DNS config

When user desires:

```yml
---
dns-resolver:
  config:
    search: []
```

It means user want to remove all search but preserve servers and
options, current nmstate incorrectly treat this as purge also.

This patch only treat these two as purge.

```yml
dns-resolver:
  config: {}
```

and

```yml
dns-resolver:
  config:
    server: []
    search: []
    options: []
```

Unit and integration test case included.

Signed-off-by: Gris Ge <fge@redhat.com>

ffmancera reviewed (approved):
LGTM!
